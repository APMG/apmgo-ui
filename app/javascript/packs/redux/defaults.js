import { PlaylistItemType, AudioPlayerType } from './types'

export class DefaultState {
  data: Array<PlaylistItemType> = []
  audioPlayers: Array<AudioPlayerType> = []
  errorMessage: string = ''
}

export class ActionType {
  type: string = 'DEFAULT'
  data: Array<{}> = []
  receivedAt: Number
  item_id : number
  message: string
  item: PlaylistItemType
}