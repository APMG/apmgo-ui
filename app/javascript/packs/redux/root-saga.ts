import { all, takeEvery } from "redux-saga/effects"
import { 
  INITIALIZE_PLAYLIST, 
  REMOVE_PLAYLIST_ITEM, 
  initializePlaylistItemsSaga, 
  removePlaylistItemSaga 
} from "./playlist"

// Take all the individual sagas and compose them into one root saga to bootstrap onto the saga middleware
export default function* rootSaga() {
  yield all([
    takeEvery(INITIALIZE_PLAYLIST, initializePlaylistItemsSaga),
    takeEvery(REMOVE_PLAYLIST_ITEM, removePlaylistItemSaga)
  ])
}