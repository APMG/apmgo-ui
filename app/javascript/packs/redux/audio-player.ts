import { ActionType, DefaultState } from './defaults';
import { PlaylistItemType } from './types';
import AudioPlayerModel from '../models/AudioPlayerModel';

export const PLAY_AUDIO_PLAYER = 'PLAY_AUDIO_PLAYER'
export const PAUSE_AUDIO_PLAYER = 'PAUSE_AUDIO_PLAYER'

export default function reducer(state : Array<AudioPlayerModel> = [], action : ActionType = new ActionType) {

  switch(action.type) {

    case PLAY_AUDIO_PLAYER:
      // append the new player if it doesn't already exist
      if (!state.find(player => player.item_id === action.item_id)) {
        state.push(new AudioPlayerModel({item_id: action.item_id}))
      }

      return state.map((player: AudioPlayerModel) => {
          // play the item being played and pause all the others
          let paused = player.item_id != action.item_id
          return player.setPaused(paused)
      })

    case PAUSE_AUDIO_PLAYER: 
      return state.map((player: AudioPlayerModel )=> {
        if (player.item_id === action.item_id) {
          return player.setPaused(true)
        }
        return player
      })

    default:
      return state
  }
}

export function playAudioPlayer (item_id) {
  return {
    type: PLAY_AUDIO_PLAYER,
    item_id: item_id
  }
}

export function pauseAudioPlayer (item_id) {
  return {
    type: PAUSE_AUDIO_PLAYER,
    item_id: item_id
  }
}