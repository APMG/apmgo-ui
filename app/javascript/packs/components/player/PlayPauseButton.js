// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { pauseAudioPlayer, playAudioPlayer } from '../../redux/audio-player';

type PlayPauseProps = {
  audio: HTMLAudioElement,
  paused: boolean,
  item_id: number,
  play: () => {},
  pause: () => {}
}

class PlayPauseButton extends React.Component {

  props: PlayPauseProps

  componentWillReceiveProps(newProps) {
    if(newProps.paused) {
      newProps.audio.pause()
    } else {
      newProps.audio.play()
    }
  }

  render() {
    if (this.props.paused || typeof this.props.paused === 'undefined') {
      return (
        <button onClick={this.props.play}>Play</button>
      )
    } else {
      return (
        <button onClick={this.props.pause}>Pause</button>
      )
    }
  }
}

const mapStateToProps = (newState) => {
  console.log(newState)
  return {
    paused: newState.audioPlayer.paused,
    item_id: newState.audioPlayer.currentTrackId
  }
}

const mapDispatchToProps = (dispatch, ownProps: PlayPauseProps) => {
  return {
    play: () : void => {
      dispatch(playAudioPlayer(ownProps.item_id))
    },
    pause: (): void => {
      dispatch(pauseAudioPlayer(ownProps.item_id))
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PlayPauseButton)
