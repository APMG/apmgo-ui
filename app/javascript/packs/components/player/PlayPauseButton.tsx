import * as React from 'react';
import { connect } from 'react-redux';
import { pauseAudioPlayer, playAudioPlayer } from '../../redux/audio-player';
import { AudioPlayerType } from '../../redux/types';
import AudioPlayer from '../../models/AudioPlayer'
import { DefaultState } from '../../redux/defaults'

interface PlayPauseProps {
  item_id: number
  player?: AudioPlayerType
  play?: () => void
  pause?: () => void
}

class PlayPauseButton extends React.Component<PlayPauseProps, {}> {

  props: PlayPauseProps

  render() {
    if (this.props.player.paused) {
      return (
        <button onClick={this.props.play}>Play</button>
      )
    } else {
      return (
        <button onClick={this.props.pause}>Pause</button>
      )
    }
  }
}

const mapDispatchToProps = (dispatch, ownProps: PlayPauseProps) => {
  return {
    play: () : void => {
      dispatch(playAudioPlayer(ownProps.item_id))
    },
    pause: (): void => {
      dispatch(pauseAudioPlayer(ownProps.item_id))
    }
  }
}

const mapStateToProps = (state : any, ownProps: PlayPauseProps) => {
  let player = state.audioPlayers.find(player => player.item_id === ownProps.item_id),
      newPlayer = new AudioPlayer(ownProps.item_id)

  newPlayer.paused = player ? player.paused : true;

  return {
    player: newPlayer
  }
}
export default connect<{}, {}, PlayPauseProps>(
  mapStateToProps, 
  mapDispatchToProps
)(PlayPauseButton)