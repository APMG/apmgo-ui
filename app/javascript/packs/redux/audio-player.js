import { take, call, put } from 'redux-saga/effects'

import { ActionType } from './defaults'
import type { PlaylistItemType } from './types'
import AudioPlayerState from '../models/AudioPlayerState'
import { RECEIVE_PLAYLIST_ITEMS } from './playlist'

export const AUDIO_META_DATA_LOADED: string = 'AUDIO_META_DATA_LOADED'
export const AUDIO_CAN_PLAY: string = 'AUDIO_CAN_PLAY'
export const PLAY_CLICK: string = 'PLAY_CLICK'
export const PAUSE_CLICK: string = 'PAUSE_CLICK'
export const MUTE_CLICK:  string = 'MUTE_CLICK'
export const UNMUTE_CLICK: string = 'UNMUTE_CLICK'
export const VOLUME_CHANGE: string = 'VOLUME_CHANGE'
export const TIME_SCRUBBER_CHANGE: string = 'TIME_SCRUBBER_CHANGE'
export const CHANGE_TRACK: string = 'CHANGE_TRACK'
export const UPDATE_PLAYTIME: string = 'UPDATE_PLAYTIME'

const defaultProps = {
  paused: true,
  muted: false,
  canPlay: false,
  volume: .5
}
const defaultPlayer = new AudioPlayerState(defaultProps)

export default function reducer(playerState : AudioPlayerState = defaultPlayer, action : ActionType = new ActionType) {

  switch(action.type) {

    case AUDIO_META_DATA_LOADED:
      return playerState.setDuration(action.duration)

    case AUDIO_CAN_PLAY:
      let canPlayState = playerState
        .setCanPlay(true)

      return canPlayState.playWhenCan
        ? canPlayState.play()
        : canPlayState

    case PLAY_CLICK:
      return playerState.play()

    case PAUSE_CLICK:
      return playerState.pause()

    case MUTE_CLICK:
      return playerState.mute()

    case UNMUTE_CLICK:
      return playerState.unmute()

    case VOLUME_CHANGE:
      return playerState.setVolume(action.volume)

    case TIME_SCRUBBER_CHANGE:
      return playerState
        .setTime(action.currentTime)
        .setCanPlay(false)
        .instanceUpdatesAudioElementTime()

    case UPDATE_PLAYTIME:
      return playerState.setTime(action.currentTime)

    case CHANGE_TRACK:
      let newplayer = playerState
        .setCanPlay(false)  // this will change when the audio element canplay event fires
        .setPlayWhenCan(!playerState.paused) // and if the player is currently playing, it will restart on canPlay
        .setCurrentTrackId(action.item.id)
        .setTime(action.item.attributes.playtime)
        .pause()  // pause while loading new audio
        .instanceUpdatesAudioElementTime() // tell state that actual audio element needs to be updated
      return newplayer

    case RECEIVE_PLAYLIST_ITEMS:
      // we need to initialize the audio player here

      // first, try to default to the first unfinished track
      let first = action.data.find(item => item.attributes.finished !== true);

      // if that fails, just take the first track
      if (!first) {
        first = action.data[0]
      }

      // if there are no tracks, just return the player unmodified
      if (!first) {
        return playerState._make({})
      }

      // otherwise set the track
      return playerState.setCurrentTrackId(first.id).setTime(first.attributes.playtime)

    default:
      return playerState
  }
}

export function audioMetaDataLoaded(duration) {
  return {
    type: AUDIO_META_DATA_LOADED,
    duration: duration // so far this is the only param we need ...
  }
}

export function audioCanPlay() {
  return {
    type: AUDIO_CAN_PLAY
  }
}

export function playClick(item_id: number) {
  return {
    type: PLAY_CLICK,
    item_id: item_id
  }
}

export function pauseClick(item_id: number) {
  return {
    type: PAUSE_CLICK,
    item_id: item_id
  }
}

export function muteClick() {
  return {
    type: MUTE_CLICK
  }
}

export function unmuteClick() {
  return {
    type: UNMUTE_CLICK
  }
}

export function volumeChange(volume) {
  return {
    type: VOLUME_CHANGE,
    volume: volume
  }
}

export function timeScrubberChange(item_id: number, currentTime: number) {
  return {
    type: TIME_SCRUBBER_CHANGE,
    item_id: item_id,
    currentTime: currentTime
  }
}

export function updatePlayTime(item_id: number, currentTime: number) {
  return {
    type: UPDATE_PLAYTIME,
    item_id: item_id,
    currentTime: currentTime
  }
}

export function changeTrack(item: PlaylistItemType) {
  return {
    type: CHANGE_TRACK,
    item: item
  }
}
