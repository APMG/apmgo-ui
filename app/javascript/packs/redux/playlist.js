import type { PlaylistItemType } from './types';
import { addItemToPlaylist, fetchPlaylistItems, deletePlaylistItem, apiArchivePlaylistItem } from '../service/playlist'
import { ActionType } from './defaults'
import { PAUSE_CLICK, UPDATE_PLAYTIME } from './audio-player'

import { put, takeLatest, call } from 'redux-saga/effects'

// Actions
export const RECEIVE_PLAYLIST_ITEMS : string = 'RECEIVE_PLAYLIST_ITEMS'
export const INITIALIZE_PLAYLIST : string = 'INITIALIZE_PLAYLIST'
export const FETCHING_PLAYLIST_ITEMS : string = 'FETCHING_PLAYLIST_ITEMS'
export const ITEM_ADDED_TO_PLAYLIST : string = 'ITEM_ADDED_TO_PLAYLIST'

export const REMOVING_PLAYLIST_ITEM : string = 'REMOVING_ITEM_FROM_PLAYLIST'
export const REMOVE_PLAYLIST_ITEM : string = 'REMOVE_PLAYLIST_ITEM'
export const PLAYLIST_ITEM_REMOVED : string = 'PLAYLIST_ITEM_REMOVED'

export const ARCHIVING_PLAYLIST_ITEM : string = 'ARCHIVING_PLAYLIST_ITEM'
export const ARCHIVE_PLAYLIST_ITEM : string = 'ARCHIVE_PLAYLIST_ITEM'
export const PLAYLIST_ITEM_ARCHIVED : string = 'PLAYLIST_ITEM_ARCHIVED'

export const PLAYLIST_ERROR_OCCURRED : string = 'PLAYLIST_ERROR_OCCURED'
export const CLEAR_PLAYLIST_ERROR : string = 'CLEAR_PLAYLIST_ERROR'

export const UPDATE_PLAYLIST_ITEM : string = 'UPDATE_PLAYLIST_ITEM'

// Statuses
class PlaylistStatuses {
  ARCHIVING_ITEM : string = 'ARCHIVING_ITEM'
  REMOVING_ITEM : string = 'REMOVING_ITEM'
  FETCHING : string = 'FETCHING'
  DEFAULT : string = 'DEFAULT'
  ERROR : string = 'ERROR'
}

// Reducer
export default function reducer(state : {data: Array<PlaylistItemType>, errorMessage: string} = {data: [], errorMessage: ''}, action : ActionType = new ActionType) {
  switch (action.type) {
    case RECEIVE_PLAYLIST_ITEMS:
      return Object.assign({}, state, {
        data: action.data,
        receivedAt: action.receivedAt,
        status: PlaylistStatuses.DEFAULT
      })

    case FETCHING_PLAYLIST_ITEMS:
      return Object.assign({}, state, {
        status: PlaylistStatuses.FETCHING,
        receivedAt: action.receivedAt
      })

    case REMOVING_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        status: PlaylistStatuses.REMOVING_ITEM
      })

    case REMOVE_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        data: state.data.filter(item => {
          return item.id !== action.item_id
        })
      })

    case PLAYLIST_ITEM_REMOVED:
      return Object.assign({}, state, {
        status: PlaylistStatuses.DEFAULT
      })

    case ARCHIVING_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        status: PlaylistStatuses.ARCHIVING_ITEM
      })

    case ARCHIVE_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        data: state.data.map(item => {
          if (item.id === action.item.id) {
            return Object.assign({}, item, action.item)
          } else {
            return item
          }
        })
      })

    case PLAYLIST_ITEM_ARCHIVED:
      return Object.assign({}, state, {
        status: PlaylistStatuses.DEFAULT
      })

    case PLAYLIST_ERROR_OCCURRED:
      return Object.assign({}, state, {
        errorMessage: action.message
      })

    case CLEAR_PLAYLIST_ERROR:
      let result = Object.assign({}, state);
      delete result.errorMessage;
      return result

    case UPDATE_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        data: state.data.map(item => {
          if (item.id === action.item.id) {
            return Object.assign({}, item, action.item)
          }
          return item;
        })
      })

    case UPDATE_PLAYTIME:
      let updated = Object.assign({}, state, {
        data: state.data.map(item => {
          if (item.id === action.item_id) {
            item.attributes.playtime = action.currentTime
            return {...item}
          } else {
            return item
          }
        })
      })
      return updated

    default: return state
  }
}

// Action creators
export function receivePlaylistItems (data) {
  return {
    type: RECEIVE_PLAYLIST_ITEMS,
    data: data,
    receivedAt: Date.now()
  }
}


export function removingPlaylistItem(item_id) {
  return {
    type: REMOVING_PLAYLIST_ITEM,
    item_id: item_id,
    receivedAt: Date.now()
  }
}

export function removePlaylistItem (access_token, item_id) {
  return {
    type: REMOVE_PLAYLIST_ITEM,
    access_token: access_token,
    item_id: item_id,
    receivedAt: Date.now()
  }
}

export function playlistItemRemoved(item_id) {
  return {
    type: PLAYLIST_ITEM_REMOVED,
    item_id: item_id,
    receivedAt: Date.now()
  }
}

// Archiving
export function archivingPlaylistItem(item) {
  return {
    type: ARCHIVING_PLAYLIST_ITEM,
    item: item,
  }
}

export function archivePlaylistItem (access_token, item) {
  return {
    type: ARCHIVE_PLAYLIST_ITEM,
    access_token: access_token,
    item: item
  }
}

export function playlistItemArchived(item) {
  return {
    type: PLAYLIST_ITEM_ARCHIVED,
    item: item
  }
}

export function addedItemToPlaylist(json) {
  return {
    type: ITEM_ADDED_TO_PLAYLIST,
    data: json.data,
    receivedAt: Date.now()
  }
}

export function initializePlaylist(access_token) {
  return {
    type: INITIALIZE_PLAYLIST,
    access_token: access_token,
    receivedAt: Date.now()
  }
}

export function fetchingPlaylistItems() {
  return {
    type: FETCHING_PLAYLIST_ITEMS,
    receivedAt: Date.now()
  }
}

export function playlistErrorOccured(message) {
  return {
    type: PLAYLIST_ERROR_OCCURRED,
    message: message,
    receivedAt: Date.now()
  }
}

export function clearPlaylistError() {
  return {
    type: CLEAR_PLAYLIST_ERROR,
    receivedAt: Date.now()
  }
}

export function updatePlaylistItem(item) {
  return {
    type: UPDATE_PLAYLIST_ITEM,
    item: item
  }
}

// Sagas

export function* initializePlaylistItemsSaga(action) {
  // Dispatch the "FETCHING_PLAYLIST_ITEMS" action to update the application status
  yield put( fetchingPlaylistItems() )

  try {
    // call async fetchPlaylistItems api function.
    // when the promise returns, yield that value from the generator
    // and also store its value to the playlist variable
    // (noted bc it was confusing to me that both of those things would happen)
    let playlist = yield call( fetchPlaylistItems, action.access_token )

    // Dispatch the RECEIVE_PLAYLIST_ITEMS action with the data from the prior call
    yield put( receivePlaylistItems(playlist) );
  } catch (e) {
    yield put( playlistErrorOccured(e.message) ) ;
  }
}

export function* removePlaylistItemSaga(action) {
  yield put ( removingPlaylistItem(action.item_id) );

  try {
    yield call( deletePlaylistItem, action.access_token, action.item_id )
    yield put ( playlistItemRemoved( action.item_id ) )
  } catch (e) {
    yield put( playlistErrorOccured( e.message ) )
  }
}

export function* archivePlaylistItemSaga(action) {
  yield put ( archivingPlaylistItem(action.item) );

  try {
    let itemResult = yield call( apiArchivePlaylistItem, action.access_token, action.item )
    yield put ( playlistItemArchived( itemResult ) )
  } catch (e) {
    yield put( playlistErrorOccured( e.message ) )
  }
}
