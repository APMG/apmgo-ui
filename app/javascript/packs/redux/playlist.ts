import { PlaylistItemType } from './types';
import { addItemToPlaylist, fetchPlaylistItems, deletePlaylistItem } from '../service/playlist'

import { put, takeLatest, call } from 'redux-saga/effects'

// Actions
export const RECEIVE_PLAYLIST_ITEMS = 'RECEIVE_PLAYLIST_ITEMS'
export const INITIALIZE_PLAYLIST = 'INITIALIZE_PLAYLIST'
export const FETCHING_PLAYLIST_ITEMS = 'FETCHING_PLAYLIST_ITEMS'
export const ITEM_ADDED_TO_PLAYLIST = 'ITEM_ADDED_TO_PLAYLIST'

export const REMOVING_PLAYLIST_ITEM = 'REMOVING_ITEM_FROM_PLAYLIST'
export const REMOVE_PLAYLIST_ITEM = 'REMOVE_PLAYLIST_ITEM'
export const PLAYLIST_ITEM_REMOVED = 'PLAYLIST_ITEM_REMOVED'

export const PLAYLIST_ERROR_OCCURRED = 'PLAYLIST_ERROR_OCCURED'
export const CLEAR_PLAYLIST_ERROR = 'CLEAR_PLAYLIST_ERROR'

// Statuses
class PlaylistStatuses {
  static readonly REMOVING_ITEM = 'REMOVING_ITEM'
  static readonly FETCHING = 'FETCHING'
  static readonly DEFAULT = 'DEFAULT'
  static readonly ERROR = 'ERROR'
}

// Reducer
export default function reducer(state = { data: [], errorMessage: '' }, action = { type: 'DEFAULT', data: {}, receivedAt: Date.now(), item_id: null, message: '' }) {
  switch (action.type) {
    case RECEIVE_PLAYLIST_ITEMS:
      return Object.assign({}, state, {
        data: action.data,
        receivedAt: action.receivedAt,
        status: PlaylistStatuses.DEFAULT
      })

    case FETCHING_PLAYLIST_ITEMS: 
      return Object.assign({}, state, {
        status: PlaylistStatuses.FETCHING,
        receivedAt: action.receivedAt
      })

    case REMOVING_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        status: PlaylistStatuses.REMOVING_ITEM
      })

    case REMOVE_PLAYLIST_ITEM:
      return Object.assign({}, state, {
        data: state.data.filter(item => {
          return item.id !== action.item_id
        })
      })

    case PLAYLIST_ITEM_REMOVED:
      return Object.assign({}, state, {
        status: PlaylistStatuses.DEFAULT
      })

    case PLAYLIST_ERROR_OCCURRED: 
      return Object.assign({}, state, {
        errorMessage: action.message
      })
    
    case CLEAR_PLAYLIST_ERROR:
      let result = Object.assign({}, state);
      delete result.errorMessage;
      return result;

    default: return state
  }
}

// Action creators
export function receivePlaylistItems (data) {
  return {
    type: RECEIVE_PLAYLIST_ITEMS,
    data: data,
    receivedAt: Date.now()
  }
}


export function removingPlaylistItem(item_id) {
  return {
    type: REMOVING_PLAYLIST_ITEM,
    item_id: item_id,
    receivedAt: Date.now()
  }
}

export function removePlaylistItem (access_token, item_id) {
  return {
    type: REMOVE_PLAYLIST_ITEM,
    access_token: access_token,    
    item_id: item_id,
    receivedAt: Date.now()
  }
}

export function playlistItemRemoved(item_id) {
  return {
    type: PLAYLIST_ITEM_REMOVED,
    item_id: item_id,
    receivedAt: Date.now()
  }
}

export function addedItemToPlaylist(json) {
  return {
    type: ITEM_ADDED_TO_PLAYLIST,
    data: json.data,
    receivedAt: Date.now()
  }
}

export function initializePlaylist(access_token) {
  return {
    type: INITIALIZE_PLAYLIST,
    access_token: access_token,
    receivedAt: Date.now()
  }
}

export function fetchingPlaylistItems() {
  return {
    type: FETCHING_PLAYLIST_ITEMS,
    receivedAt: Date.now()
  }
}

export function playlistErrorOccured(message) {
  return {
    type: PLAYLIST_ERROR_OCCURRED,
    message: message,
    receivedAt: Date.now()
  }
}

export function clearPlaylistError() {
  return {
    type: CLEAR_PLAYLIST_ERROR,
    receivedAt: Date.now()
  }
}

// Sagas 

export function* initializePlaylistItemsSaga(action) {
  // Dispatch the "FETCHING_PLAYLIST_ITEMS" action to update the application status
  yield put( fetchingPlaylistItems() )

  try {
    // call async fetchPlaylistItems api function.
    // when the promise returns, yield that value from the generator 
    // and also store its value to the playlist variable 
    // (noted bc it was confusing to me that both of those things would happen) 
    let playlist = yield call( fetchPlaylistItems, action.access_token )
    // TODO: wrap this ^^^ in a try/catch block to handle errors

    // Dispatch the RECEIVE_PLAYLIST_ITEMS action with the data from the prior call
    yield put( receivePlaylistItems(playlist) );
  } catch (e) {
    yield put( playlistErrorOccured(e.message) ) ;
  }
}

export function* removePlaylistItemSaga(action) {
  yield put ( removingPlaylistItem(action.item_id) );

  try {
    yield call( deletePlaylistItem, action.access_token, action.item_id )
    yield put ( playlistItemRemoved( action.item_id ) )
  } catch (e) {
    yield put( playlistErrorOccured( e.message ) )
  }

}