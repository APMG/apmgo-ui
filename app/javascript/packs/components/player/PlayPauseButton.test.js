
import React from 'react'
import { mount } from 'enzyme'
import { getSnapshotJson, getWrappedComponent, itemFixtures } from '../../redux/__tests__/mock-initial-state'
import { playClick, pauseClick } from '../../redux/audio-player'

import PlayPauseButton from './PlayPauseButton'

describe('Play/Pause Button Component', () => {
  describe('Presenter', () => {
    it('Displays a Loading Button When It Can Not Play', () => {
      let tree = getSnapshotJson(<PlayPauseButton canPlay={false}/>)
      expect(tree).toMatchSnapshot()
    })
    it('Displays a Play Button When Paused', () => {
      let props = {
        canPlay: true,
        paused: true
      },
      tree = getSnapshotJson(<PlayPauseButton {...props} />)
      expect(tree).toMatchSnapshot()
    })
    it('Displays a Pause Button When Playing', () => {
      let props = {
        canPlay: true,
        paused: true
      },
      tree = getSnapshotJson(<PlayPauseButton {...props} />)
      expect(tree).toMatchSnapshot()
    })
    it('Calls `play` When Play Button Is Clicked', () => {
      let props = {
        canPlay: true,
        paused: true,
        play: jest.fn()
      },
      component = getWrappedComponent(<PlayPauseButton {...props} />),
      wrapper = mount(component)

      wrapper.find('#play-button').simulate('click')
      expect(props.play.mock.calls.length).toEqual(1);
    })
    it('Calls `pause` When Pause Button Is Clicked', () => {
      let props = {
        canPlay: true,
        paused: false,
        pause: jest.fn()
      },
      component = getWrappedComponent(<PlayPauseButton {...props} />),
      wrapper = mount(component)

      wrapper.find('#pause-button').simulate('click')
      expect(props.pause.mock.calls.length).toEqual(1);
    })
  })
})
