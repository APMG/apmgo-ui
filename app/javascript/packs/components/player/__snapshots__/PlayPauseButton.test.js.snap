// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play/Pause Button Component Presenter Displays a Loading Button When It Can Not Play 1`] = `
<button
  disabled={true}
  id="play-pause-loading-button"
>
  Loading ...
</button>
`;

exports[`Play/Pause Button Component Presenter Displays a Pause Button When Playing 1`] = `
<button
  id="play-button"
  onClick={undefined}
>
  Play
</button>
`;

exports[`Play/Pause Button Component Presenter Displays a Play Button When Paused 1`] = `
<button
  id="play-button"
  onClick={undefined}
>
  Play
</button>
`;
